### PR TITLE
Query Analysis

### PR DESCRIPTION
**Pull request by nishchay-veer**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

• The QueryAPI now includes a parameter called `analyze`.  When set to true, the `Analyze()` method of the query returns query analysis. This analysis is represented by the `AnalyzeOutputNode` structure, as mentioned in [#289](https://github.com/thanos-io/promql-engine/pull/289/files), and it is included in the response.

## Screenshots
Screenshot from 2023-09-07 20-57-45

Screenshot from 2023-08-17 13-10-30
Screenshot from 2023-08-21 01-15-56

## Verification

Tested locally

### CODE REVIEW

**Comment by saswatamcode** [comment]
Let's update promql-engine here with your commit now 🙂 `go get github.com/thanos-io/promql-engine@99d2be3bb060450a6d2a039a28d9562876c4c204`

File: pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx [hunk 1]
```diff
@@ -89,10 +93,30 @@ const ExpressionInput: FC<PathPrefixProps & CMExpressionInputProps> = ({
...
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({ query: value }),
+    })
+      .then((response) => response.json())
+      .then((json) => {
+        if (json.status === 'success') {
+          setExplainOutput(json.data);
+        } else {
+          // Handle error here
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx:113]: Let's actually handle the err here

File: pkg/api/query/v1_test.go [hunk 2]
```diff
@@ -628,6 +629,70 @@ func TestQueryEndpoints(t *testing.T) {
...
+		queryableCreate:       query.NewQueryableCreator(nil, nil, newProxyStoreWithTSDBStore(db), 2, timeout),
+		engineFactory:         ef,
+		defaultEngine:         PromqlEnginePrometheus,
+		lookbackDeltaCreate:   func(m int64) time.Duration { return time.Duration(0) },
+		gate:                  gate.New(nil, 4, gate.Queries),
+		defaultRangeQueryStep: time.Second,
+		queryRangeHist: promauto.With(prometheus.NewRegistry()).NewHistogram(prometheus.HistogramOpts{
+			Name: "query_range_hist",
+		}),
+		seriesStatsAggregator: &store.NoopSeriesStatsAggregator{},
```
Comment (saswatamcode) [inline pkg/api/query/v1_test.go:658]: Hmm this should be `seriesStatsAggregatorFactory` right?

**Review by saswatamcode — COMMENTED** [review]
Could you attach some screenshots of the Query UI in the PR desc? :)

File: pkg/api/query/v1.go [hunk 3]
```diff
@@ -78,6 +78,7 @@ const (
 	LookbackDeltaParam       = "lookback_delta"
 	EngineParam              = "engine"
 	QueryExplainParam        = "explain"
```
Comment (PradyumnaKrishna) [inline pkg/api/query/v1.go:80]: I guess we don't need this param now.

File: pkg/api/query/v1.go [hunk 4]
```diff
@@ -545,7 +670,7 @@ func (qapi *QueryAPI) query(r *http.Request) (interface{}, []error, *api.ApiErro
 		return nil, nil, &api.ApiError{Typ: api.ErrorBadData, Err: err}, func() {}
 	}
 
-	explanation, apiErr := qapi.parseQueryExplainParam(r, qry)
+	analysis, err := qapi.parseQueryAnalyzeParam(r, qry)
 	if apiErr != nil {
```
Comment (PradyumnaKrishna) [inline pkg/api/query/v1.go:674]: `err` should be checked not `apiErr`.

File: pkg/api/query/v1.go [hunk 5]
```diff
@@ -718,7 +975,7 @@ func (qapi *QueryAPI) queryRange(r *http.Request) (interface{}, []error, *api.Ap
...
```
Comment (PradyumnaKrishna) [inline pkg/api/query/v1.go:979]: Same as above, `err` should be checked not `apiErr`.

File: pkg/api/query/v1_test.go [hunk 6]
```diff
@@ -628,6 +638,138 @@ func TestQueryEndpoints(t *testing.T) {
...
+			},
+			response: &queryData{
+				QueryAnalysis: queryTelemetry{
+					OperatorName: "[*numberLiteralSelector] 2",
+				},
+			},
+		},
+	}
+	for i, test := range tests {
+		if ok := testEndpoint(t, test, fmt.Sprintf("#%d %s", i, test.query.Encode()), reflect.DeepEqual); !ok {
```
[the remark this answers is not in the document]
Comment (nishchay-veer) [reply pkg/api/query/v1_test.go:761]: Do I need to pass the argument in reflect.deepequal.go as well then ?

File: pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx [hunk 7]
```diff
@@ -88,11 +93,36 @@ const ExpressionInput: FC<PathPrefixProps & CMExpressionInputProps> = ({
...
 }) => {
   const containerRef = useRef<HTMLDivElement>(null);
   const viewRef = useRef<EditorView | null>(null);
   const { theme } = useTheme();
-
+  const handleExplain = async (query: string) => {
+    try {
+      const queryParams = new URLSearchParams({
+        query,
+        engine: 'thanos',
```
[the remark this answers is not in the document]
Comment (saswatamcode) [reply pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx:107]: Do you mean that the `Explain` button should be disabled when `prometheus` engine is selected? Then I agree 🙂

File: pkg/api/query/v1.go [hunk 8]
```diff
@@ -427,25 +440,136 @@ func (qapi *QueryAPI) parseShardInfo(r *http.Request) (*storepb.ShardInfo, *api.
...
 		}
-		if enableExplanation {
-			if eq, ok := query.(engine.ExplainableQuery); ok {
-				explanation = eq.Explain()
-			} else {
-				return explanation, &api.ApiError{Typ: api.ErrorBadData, Err: errors.Errorf("Query not explainable")}
-			}
+	}
+	return queryTelemetry{}, errors.Errorf("Query not analyzable; change engine to 'thanos' ")
+}
```
Comment (saswatamcode) [inline pkg/api/query/v1.go:458]: Seems like we are always including an `analyze` param via UI, so this no longer works and always errors. So let's change the logic a bit.
```suggestion
func (qapi *QueryAPI) parseQueryAnalyzeParam(r *http.Request, query promql.Query) (queryTelemetry, error) {
	if r.FormValue(QueryAnalyzeParam) == "true" || r.FormValue(QueryAnalyzeParam) == "1" {
		if eq, ok := query.(engine.ExplainableQuery); ok {
			return processAnalysis(eq.Analyze()), nil
		}

		return queryTelemetry{}, errors.Errorf("Query not analyzable; change engine to 'thanos'")
	}

	return queryTelemetry{}, nil
}
```

File: pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx [hunk 9]
```diff
@@ -88,11 +94,39 @@ const ExpressionInput: FC<PathPrefixProps & CMExpressionInputProps> = ({
...
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx:107]: I don't think this is correct. We need to pass the same parameters as query endpoints, to the explain endpoints.

I think the best way would be to do similar to existing. So passing a function as a prop to `ExpressionInput` component called `executeExplain` which would call the explain endpoints and pass the same params as query. The `executeExplain` function can be defined in `Panel.tsx` in the same format. 🙂

**Review by saswatamcode — COMMENTED** [review]
Nice progress! We're close to finishing this, let's address a few bugs. 💪🏻

File: pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (PradyumnaKrishna) [reply pkg/ui/react-app/src/pages/graph/ExpressionInput.tsx:107]: Yes.

File: pkg/ui/react-app/src/pages/graph/Panel.tsx [hunk 10]
```diff
@@ -429,23 +514,42 @@ class Panel extends Component<PanelProps & PathPrefixProps, PanelState> {
                 <option value="thanos">Thanos</option>
               </Input>
             </div>
-            <div className="float-right">
+            <div className="float-right" onMouseEnter={this.handleMouseEnter} onMouseLeave={this.handleMouseLeave}>
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/pages/graph/Panel.tsx:517]: Got a screenshot of what on/off hover here looks like? 🙂

Comment (nishchay-veer) [reply pkg/ui/react-app/src/pages/graph/Panel.tsx:517]: Yes, included in PR desc :)

**Comment by GiedriusS** [comment]
Seems like UI tests care failing: https://app.circleci.com/pipelines/github/thanos-io/thanos?branch=pull%2F6515. @nishchay-veer could you take a look?

**Comment by GiedriusS** [comment]
`TestQFEEngineExplanation/passes_range_query_to_Thanos_engine_and_returns_explanation` fails. The test probably needs to be updated to use the new API? @nishchay-veer

File: pkg/ui/react-app/src/pages/graph/Panel.tsx [hunk 11]
```diff
@@ -429,23 +514,42 @@ class Panel extends Component<PanelProps & PathPrefixProps, PanelState> {
...
+                id={`analyze-${this.props.id}`}
+                onChange={this.handleChangeAnalyze}
+                checked={options.analyze}
+                disabled={options.disableAnalyzeCheckbox}
+                className="analyze-checkbox"
               >
-                Explain
+                Analyze
               </Checkbox>
+              {/* <Tooltip
```
Comment (GiedriusS) [inline pkg/ui/react-app/src/pages/graph/Panel.tsx:528]: Why is this commented out?

**Review by GiedriusS — COMMENTED** [review]
Maybe we could hide the Explain output if Execute is called? Because right now the Explain output is sticky and is always there if the user clicks "Explain" :smile:

File: pkg/ui/react-app/src/pages/graph/Panel.tsx (hunk 11 above)
Comment (nishchay-veer) [reply pkg/ui/react-app/src/pages/graph/Panel.tsx:528]: was just testing where the UI tests are failing, will uncomment it

Comment (nishchay-veer) [reply pkg/ui/react-app/src/pages/graph/Panel.tsx]: added customized tooltip instead of the component, the UI tests were failing due to "target" attribute

**Comment by GiedriusS** [comment]
@nishchay-veer linters are failing, could you please remove the now unused return value?

**Comment by GiedriusS** [comment]
@nishchay-veer could you also please fix the conflicts? :pray:

**Comment by GiedriusS** [comment]
@nishchay-veer I think your local `main` branch was stale and your last commit just nuked the latest updates :sweat:

**Comment by nishchay-veer** [comment]
Not working with `make assets` :(

**Review by saswatamcode — DISMISSED** [review]
LGTM, let's just address conflicts and get this merged!

**Pull request merged**